create_layout.py

```diff
@@ -91,48 +91,123 @@
 import in_toto.models.link
 import in_toto.models.layout
 
-def create_material_rules(links, index):
-  """Create generic material rules (3 variants)
+def changes_between_snapshots(before_dict, after_dict):
+  """Given two 'snapshots' of an artifacts structure -- 'before' and 'after' --
+  return a tuple specifying which artifacts have been added, which have been
+  removed, which have been modified, and which have remained unchanged. Both
+  these dictionaries have artifact names as the keys and their hashes as the
+  values."""
+
+  before_set = set(before_dict.keys())
+  after_set = set(after_dict.keys())
+
+  removed_artifacts = before_set.difference(after_set)
+  added_artifacts = after_set.difference(before_set)
+
+  unchanged_artifacts = set()
+  modified_artifacts = set()
+  for key in before_set.intersection(after_set):
+    if before_dict[key] == after_dict[key]:
+      unchanged_artifacts.add(key)
+    else:
+      modified_artifacts.add(key)
+
+  return (sorted(unchanged_artifacts), sorted(modified_artifacts), sorted(added_artifacts),
+      sorted(removed_artifacts))
+
+def create_material_rules(previous_link, current_link):
+  """Create generic material rules
+
+  - MATCH available materials with products from previous step (links must be
+      an ordered list) and
+  - ALLOW available materials if it is the first step in the list
+  - DELETE removed materials
+
+  Args:
+    previous_link: a link of previous step, including previous step's materials
+        and products
+    current link: a link of current step, including current step's materials
+        and products
+
+  Returns:
+    a list of material rules
+  """
+
+  expected_materials_rules = []
+  expected_allow_rules = []
+  expected_match_rules = []
+  expected_delete_rules = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = set(unchanged_artifacts + modified_artifacts \
+       + deleted_artifacts)
+  if previous_link is None:
+    # If current step is the initial step, we don't need to add MATCH rules.
+    # Instead, ALLOW all the existing artifacts except deleted artifact
+    for artifact in curr_step_materials.difference(deleted_artifacts):
+        expected_allow_rules.append(["ALLOW", artifact])
 
-  * MATCH available materials with products from previous step (links must be an
-  ordered list) and
-  * ALLOW available materials if it is the first step in the
-  list
-  Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  else:
+    # MATCH only those that already were in the previous step
+    # WARNING: if we have a MATCH rule and a DELETE rule on the same artifact,
```

Comment:
Oh, I meant an actual warning using e.g. the `logging` module here, not just a comment. :)